### ISSUE TITLE
AVS-SDK does not work properly with bluez-alsa

### ISSUE THREAD

**Issue by subjectxbj** [alexa/avs-device-sdk#1149]
SDK: 1.10

I'm trying to stream audio from AVS client to a bluetooth speaker via bluetooth. I have integrated bluez-alsa into my device. I can stream a local wav file to the bluetooth speaker via aplay or gstreamer without any problem.

But when I try to stream the music to the bluetooth speaker via AVS client, it just does not work. Here is the log when AVS-client receive a "Play" directive:

```
########################################
#       Alexa is currently idle!       #
########################################
2019-01-02 11:12:39.134 [ 15] I DirectiveSequencer:onDirective:directive={"namespace\:"AudioPlayer"\,name\:"Play"\,messageId\:"8d80c7a9-8549-40f2-8621-4d4cd74f0631"\,dialogRequestId\:""}
2019-01-02 11:12:39.135 [ 16] I DirectiveRouter:preHandleDirective:messageId=8d80c7a9-8549-40f2-8621-4d4cd74f0631,action=calling
2019-01-02 11:12:39.137 [ 16] 0 CapabilityAgent:addingMessageIdToMap:messageId=8d80c7a9-8549-40f2-8621-4d4cd74f0631
2019-01-02 11:12:39.138 [ 17] I DirectiveRouter:handleDirective:messageId=8d80c7a9-8549-40f2-8621-4d4cd74f0631,action=calling
2019-01-02 11:12:39.139 [ 17] 0 AudioPlayer:handleDirective:name=Play,messageId=8d80c7a9-8549-40f2-8621-4d4cd74f0631
2019-01-02 11:12:39.141 [ 17] 1 AudioPlayer:handlePlayDirective
2019-01-02 11:12:39.142 [ 17] E JsonUtils:findNodeFailed:reason=missingDirectChild,child=streamFormat
2019-01-02 11:12:39.143 [ 17] E JsonUtils:findNodeFailed:reason=missingDirectChild,child=progressReport
2019-01-02 11:12:39.144 [ 17] E JsonUtils:findNodeFailed:reason=missingDirectChild,child=expectedPreviousToken
2019-01-02 11:12:39.145 [ 17] 0 DirectiveProcessor:onHandlingCompeted:messageId=8d80c7a9-8549-40f2-8621-4d4cd74f0631,directiveBeingPreHandled=(nullptr)
2019-01-02 11:12:39.146 [ 17] 0 CapabilityAgent:removingMessageIdFromMap:messageId=8d80c7a9-8549-40f2-8621-4d4cd74f0631
2019-01-02 11:12:39.148 [  f] 1 AudioPlayer:executePlay:playBehavior=REPLACE_ALL
2019-01-02 11:12:39.149 [  f] 1 AudioPlayer:executeStop:playNextItem=true,m_currentActivity=IDLE
2019-01-02 11:12:39.150 [  f] 1 FocusManager:acquireChannel:channelName=Content,interface=AudioPlayer
2019-01-02 11:12:39.151 [ 18] 0 AudioPlayer:onFocusChanged:newFocus=FOREGROUND
2019-01-02 11:12:39.152 [  f] 1 AudioPlayer:executeOnFocusChanged:from=NONE,to=FOREGROUND,m_currentActivity=IDLE
2019-01-02 11:12:39.154 [  f] 1 AudioPlayer:executeOnFocusChanged:action=playNextItem
2019-01-02 11:12:39.155 [  f] 1 AudioPlayer:playNextItem:m_audioItems.size=1
2019-01-02 11:12:39.156 [  3] 0 MediaPlayer:handleSetSourceForUrlCalled
2019-01-02 11:12:39.157 [  3] E MediaPlayer:saveOffsetBeforeTeardown - gst_element_query_position failed
2019-01-02 11:12:39.160 [  3] 0 MediaPlayer:handleSetAttachmentReaderSourceCalled
2019-01-02 11:12:39.162 [  3] E MediaPlayer:saveOffsetBeforeTeardown - gst_element_query_position failed
2019-01-02 11:12:39.172 [  3] 0 MediaPlayer:handlePlayCalled:idPassed=1,currentId=1
../../../src/asound/../shared/ctl-client.c:107: Connecting to socket: /var/run/bluealsa/hci0
../../../src/asound/../shared/ctl-client.c:237: Getting transport for 00:12:6F:66:98:B7 type 0x41
../../../src/asound/bluealsa-pcm.c:545: Setting constraints: 52
2019-01-02 11:12:39.223 [  3] 0 MediaPlayer:handlePlay:startingState=PAUSED,stateReturn=ASYNC
2019-01-02 11:12:40.838 [ 19] 0 InProcessAttachmentReader:readFailed:reason=SDS is closed
2019-01-02 11:12:42.553 [ 19] 0 InProcessAttachmentReader:readFailed:reason=SDS is closed
../../../src/asound/bluealsa-pcm.c:293: Initializing HW: 52
../../../src/asound/../shared/ctl-client.c:396: Requesting PCM open for 00:12:6F:66:98:B7
2019-01-02 11:12:43.577 [  7] E Bluetooth:onEventFired:reason=mismatchedDevices,eventDevice=00\:12\:6F\:66\:98\:B7,activeDevice=null
../../../src/asound/bluealsa-pcm.c:316: FIFO buffer size: 4096
../../../src/asound/bluealsa-pcm.c:322: Selected HW buffer: 20 periods x 1920 bytes == 38400 bytes
../../../src/asound/bluealsa-pcm.c:337: Initializing SW: 52
../../../src/asound/bluealsa-pcm.c:337: Initializing SW: 52
../../../src/asound/bluealsa-pcm.c:337: Initializing SW: 52
../../../src/asound/bluealsa-pcm.c:337: Initializing SW: 52
../../../src/asound/bluealsa-pcm.c:353: Prepared: 52
../../../src/asound/bluealsa-pcm.c:337: Initializing SW: 52
2019-01-02 11:12:43.668 [  3] 0 MediaPlayer:callingOnTags
2019-01-02 11:12:43.670 [  3] 0 AudioPlayer:onTags:id=1
2019-01-02 11:12:43.671 [  3] 0 MediaPlayer:callingOnTags
2019-01-02 11:12:43.673 [  3] 0 AudioPlayer:onTags:id=1
2019-01-02 11:12:43.672 [  f] 1 AudioPlayer:executeOnTags:id=1
2019-01-02 11:12:43.676 [  f] 0 EventBuilder:buildJsonEventString:messageId=c8cc2db3-b107-41e6-b7e8-012b7469f519,namespace=AudioPlayer,name=StreamMetadataExtracted
2019-01-02 11:12:43.678 [  f] 1 AudioPlayer:executeOnTags:id=1
2019-01-02 11:12:43.681 [  f] 0 EventBuilder:buildJsonEventString:messageId=8aa35cb5-eea0-4cc6-9a04-b661915a3b76,namespace=AudioPlayer,name=StreamMetadataExtracted
2019-01-02 11:12:43.684 [  3] 0 MediaPlayer:seekSuccessful:offsetInMs=7776
2019-01-02 11:12:43.686 [  3] 0 MediaPlayer:callingOnTags
2019-01-02 11:12:43.688 [  3] 0 AudioPlayer:onTags:id=1
2019-01-02 11:12:43.689 [  f] 1 AudioPlayer:executeOnTags:id=1
2019-01-02 11:12:43.691 [  f] 0 EventBuilder:buildJsonEventString:messageId=5f39322b-f6a6-4cfb-97d9-ef6d6f67c62a,namespace=AudioPlayer,name=StreamMetadataExtracted
2019-01-02 11:12:43.699 [  3] 0 MediaPlayer:callingOnPlaybackStarted:currentId=1
2019-01-02 11:12:43.700 [  3] 0 AudioPlayer:onPlaybackStarted:id=1
2019-01-02 11:12:43.701 [  f] 1 AudioPlayer:executeOnPlaybackStarted:id=1
2019-01-02 11:12:43.703 [  f] 0 AudioPlayer:changeActivity:from=IDLE,to=PLAYING
2019-01-02 11:12:43.704 [  f] 0 AudioPlayer:executeProvideState:sendToken=false,stateRequestToken=0
2019-01-02 11:12:43.705 [  3] 0 MediaPlayer:handleGetOffsetCalled:idPassed=1,currentId=1
2019-01-02 11:12:43.707 [  3] 0 MediaPlayer:handleGetOffsetCalled:idPassed=1,currentId=1
../../../src/asound/bluealsa-pcm.c:225: Starting: 52
2019-01-02 11:12:43.709 [  f] 1 AudioPlayer:notifyObserver:playerActivity=PLAYING
../../../src/asound/../shared/ctl-client.c:462: Requesting PCM resume for 00:12:6F:66:98:B72019-01-02 11:12:43.712 [  f] 1 AudioPlayer:sendEventWithTokenAndOffset:eventName=PlaybackStarted,offset=567776

2019-01-02 11:12:43.715 [  f] 0 EventBuilder:buildJsonEventString:messageId=cb897042-c4e8-4271-8751-595bbb71a51c,namespace=AudioPlayer,name=PlaybackStarted
../../../src/asound/bluealsa-pcm.c:119: Starting IO loop: 55
../../../src/asound/bluealsa-pcm.c:130: IO thread paused: 4
2019-01-02 11:12:44.123 [ 12] 0 MimeResponseSink:receivedRequestId:value=x-amzn-requestid\: 0ab2cbfffe656970-00001682-00cdec7a-ef3004291291c171-2ea4e8cb-7
2019-01-02 11:12:44.125 [ 12] 0 LibcurlHTTP2Connection:streamFinished:streamId=AVSEvent-7,result=No error,CURLcode=0
2019-01-02 11:12:44.509 [ 12] 0 MimeResponseSink:receivedRequestId:value=x-amzn-requestid\: 0ab2cbfffe656970-00001682-00cdec7a-ef3004291291c171-2ea4e8cb-9
2019-01-02 11:12:44.511 [ 12] 0 LibcurlHTTP2Connection:streamFinished:streamId=AVSEvent-9,result=No error,CURLcode=0
2019-01-02 11:12:44.900 [ 12] 0 MimeResponseSink:receivedRequestId:value=x-amzn-requestid\: 0ab2cbfffe656970-00001682-00cdec7a-ef3004291291c171-2ea4e8cb-11
2019-01-02 11:12:44.902 [ 12] 0 LibcurlHTTP2Connection:streamFinished:streamId=AVSEvent-11,result=No error,CURLcode=0
2019-01-02 11:12:45.562 [ 12] 0 MimeResponseSink:receivedRequestId:value=x-amzn-requestid\: 0ab2cbfffe656970-00001682-00cdec7a-ef3004291291c171-2ea4e8cb-13
2019-01-02 11:12:45.564 [ 12] 0 LibcurlHTTP2Connection:streamFinished:streamId=AVSEvent-13,result=No error,CURLcode=0
2019-01-02 11:17:07.433 [ 10] 0 InProcessAttachmentReader:readFailed:reason=SDS is closed
2019-01-02 11:17:45.861 [ 12] 0 LibcurlHTTP2Connection:streamFinished:streamId=AVSPing-15,result=No error,CURLcode=0

```

Here is my alsa conf:
```
pcm.!default {
   type asym
   playback.pcm{
     type plug
     slave.pcm "softvol"
   }
   capture.pcm {
     type plug
     slave.pcm "hw:0,0"
   }
}
defaults.bluealsa.interface "hci0"
defaults.bluealsa.device "00:12:6F:66:98:B7"
defaults.bluealsa.profile "a2dp"
defaults.bluealsa.delay 10000

pcm.softvol {
    type softvol
    slave {
        pcm "bluealsa"
    }
    control {
        name "pcm"
        card 0
    }
}

```

**Comment by kjkh** [comment]
Hi @subjectxbj,

There is an error, `2019-01-02 11:12:43.577 [  7] E Bluetooth:onEventFired:reason=mismatchedDevices,eventDevice=00\:12\:6F\:66\:98\:B7,activeDevice=null`.

This has to do with the fact that the SDK registers its own endpoints to BlueZ. BlueALSA also does this, so the SDK runs into issues initializing because its endpoint is not the one that has priority.

Can you please install the latest version of BlueALSA, and launch it? By default, a2dp-sink should be disabled, but if not, you can explicitly disable it with the `-p` option and omitting the a2dp-sink profile.

**Comment by kjkh** [comment]
Additionally, can you ensure that the SDK is running when you connect the BT speaker?

**Comment by subjectxbj** [comment]
I tried AVS-SDK 1.6 which does not have bluetooth, and AVS Client can stream music to BT speaker via bluealsa. So there is no problem on the bluealsa. 

So regarding the conflicts of AVS-SDK 1.10 and bluealsa, is there any way to remove the conflict on the AVS SDK side?

Thanks

**Comment by visayamv** [comment]
Hi @subjectxbj , did you pair the bluetooth speaker using this instruction: https://www.amazon.com/gp/help/customer/display.html?nodeId=202011820 ?

**Comment by subjectxbj** [comment]
Yes, that's how I did.

**Comment by subjectxbj** [comment]
Has the SDK team tested the SDK with Bluealsa? I checked the AVS SDK code, seems it only registered the endpoint of a SINK object. And bluealsa only register the endpoint of a Source object if I specify "--profile=a2dp-source". So there should not be any conflict.

**Comment by subjectxbj** [comment]
[log_of_a2dp_source.txt](https://github.com/alexa/avs-device-sdk/files/2731340/log_of_a2dp_source.txt)
I attached a full log at debug level debug9 when play music to a BT speaker.
Following is the step:
1. open "alexa.amazon.com", and click "Settings->Bluetooth Devices", click the BT speaker to connect
2. open "Music, Video & Books" and select a music to play.

The attached the log start from stop 2.

In the log, I did not see the log anymore:
```
 Bluetooth:onEventFired:reason=mismatchedDevices,eventDevice=00\:12\:6F\:66\:98\:B7,activeDevice=null
```

But still did not hear any sound on the BT speaker. 

From the bluealsa log, seems the "drain" function is never called.
